apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.UnitTests/SessionPoolTests.cs

```diff
@@ -0,0 +1,443 @@
+        }
+
+        [Fact]
+        public async Task MaxActiveViolationThrows()
+        {
+            using (var pool = new SessionPool())
+            {
+                pool.Options.WaitOnResourcesExhausted = false;
+                pool.Options.MaximumActiveSessions = 2;
+                bool exceptionThrown = false;
+
+                var client = CreateMockClient();
+                await pool.CreateSessionFromPoolAsync(
+                        client.Object, s_defaultName.ProjectId,
+                        s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None)
+                    .ConfigureAwait(false);
+                await pool.CreateSessionFromPoolAsync(
+                        client.Object, s_defaultName.ProjectId,
+                        s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None)
+                    .ConfigureAwait(false);
+                try
+                {
+                    await pool.CreateSessionFromPoolAsync(
+                        client.Object, s_defaultName.ProjectId,
+                        s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None)
+                        .ConfigureAwait(false);
+                }
+                catch (RpcException e)
+                {
+                    Assert.Equal(StatusCode.ResourceExhausted, e.Status.StatusCode);
+                    exceptionThrown = true;
+                }
+
+                Assert.True(exceptionThrown);
+            }
+        }
+
+        [Fact]
+        public async Task MaxActiveViolationBlocks()
+        {
+            using (var pool = new SessionPool())
+            {
+                pool.Options.WaitOnResourcesExhausted = true;
+                pool.Options.MaximumActiveSessions = 2;
+
+                var client = CreateMockClient();
+                var session1 = await pool.CreateSessionFromPoolAsync(
+                    client.Object, s_defaultName.ProjectId,
+                    s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None)
+                    .ConfigureAwait(false);
+                await pool.CreateSessionFromPoolAsync(
+                    client.Object, s_defaultName.ProjectId,
+                    s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None)
+                    .ConfigureAwait(false);
+
+                var releaseTask = Task.Delay(TimeSpan.FromMilliseconds(10)).ContinueWith(
+                    (x, y) => { pool.ReleaseToPool(client.Object, session1); }, null);
+
+                var createTask = pool.CreateSessionFromPoolAsync(
+                    client.Object, s_defaultName.ProjectId,
+                    s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None);
+
+                await Task.WhenAll(createTask, releaseTask).ConfigureAwait(false);
+                Assert.NotNull(createTask.ResultWithUnwrappedExceptions());
```

Comment:
actually this turned out to reveal an ordering issue that while doesn't result in a bug, could result in an unnecessary session alloc -- good catch!